### PR TITLE
Avoid optional instances in Loops

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Minor refactor, part of #10324

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [n/a] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda @justusschock @awaelchli @akihironitta @carmocca @ananthsub @ninginthecloud

### CODE REVIEW

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py [hunk 1]
```diff
@@ -44,16 +44,13 @@ def __init__(self) -> None:
...
-        self._dataloader_state_dict: Dict[str, Any] = None
+        self._dataloader_state_dict: Dict[str, Any] = {}
 
     @property
     def done(self) -> bool:
         """Returns ``True`` if the current iteration count reaches the number of dataloader batches."""
         return self.batch_progress.current.completed >= self._dl_max_batches
 
-    def connect(self, **kwargs: "Loop") -> None:
-        raise NotImplementedError(f"{self.__class__.__name__} does not connect any child loops.")
```
Comment (awaelchli) [inline pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:55]: this is here intentionally! otherwise eval_loop.connect(loop) will do nothing

Comment (carmocca) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:55]: This is here accidentally.

1. No other leaf loops override `connect` with the `NotImplementedError`.

2. The base `Loop.connect` does nothing, if we wanted the `NotImplementedError` behaviour, it should be in the base.

3. Erroring out here makes #10324 harder, as we would need to check whether a loop has any children loop before calling `connect`.

Comment (awaelchli) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:55]: Fine. Remove it. With Loop.replace the connect method gets removed anyway

**Review by awaelchli — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by justusschock — APPROVED** [review]

**Review by akihironitta — APPROVED** [review]
